### PR TITLE
[5.1] Use BaseCollection for non-model returns

### PR DESCRIPTION
**Pull request by rkgrep**
Recreated #9189

Git failed to overwrite old branch for some reason

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Collection.php [hunk 1]
```diff
@@ -60,6 +60,69 @@ public function add($item)
     }
 
     /**
+     * Get an array with the values of a given key.
+     *
+     * @param  string  $value
+     * @param  string  $key
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Collection.php]: string|null

File: src/Illuminate/Database/Eloquent/Collection.php [hunk 2]
```diff
@@ -60,6 +60,69 @@ public function add($item)
...
+     */
+    public function keys()
+    {
+        return new BaseCollection(array_keys($this->items));
+    }
+
+    /**
+     * Zip the collection together with one or more arrays.
+     *
+     * @param  mixed ...$items
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Collection.php]: not correct phpdoc

File: src/Illuminate/Database/Eloquent/Collection.php [hunk 3]
```diff
@@ -60,6 +60,69 @@ public function add($item)
...
+     * @return \Illuminate\Support\Collection
+     */
+    public function zip($items)
+    {
+        return new BaseCollection(parent::zip($items));
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Collection.php]: you need func_get_args here for the dynamic stuff to work...

**Comment by rkgrep** [comment]
Also provided tests and corrected phpDoc in base collection.

File: src/Illuminate/Database/Eloquent/Collection.php [hunk 4]
```diff
@@ -60,6 +60,69 @@ public function add($item)
...
+    }
+
+    /**
+     * Flip the items in the collection.
+     *
+     * @return \Illuminate\Support\Collection
+     */
+    public function flip()
+    {
+        return new BaseCollection(array_flip($this->items));
```
Comment (tonglil) [inline src/Illuminate/Database/Eloquent/Collection.php]: I think you can use the `make()` method like `return BaseCollection::make($this->items)->flip()` to save yourself from writing implementation here?

**Comment by TommyC81** [comment]
@rkgrep Thanks for taking the time to do this!

**Comment by rkgrep** [comment]
@tonglil thanks for a hint.

**Comment by lucasmichot** [comment]
@rkgrep  I'm :-1: 
- your tests pass even the old and the updated methods
- what happens if in my code I pass this collection as an argument and take a specific action if this a `Illuminate/Database/Eloquent/Collection` ?

**Comment by tonglil** [comment]
@rkgrep Sure thing, though I think using `::make()` would look better than wrapping the constructor in pareths.

**Comment by taylorotwell** [comment]
Definitely can't make this change on the 5.1 branch, if at all... since it would be breaking as @lucasmichot noted.

**Comment by Sladewill** [comment]
What about #9163 there's already an issue from there.

**Comment by TommyC81** [comment]
Yeah, can confirm the issue. Something went wrong here, unfortunately.

**Comment by rkgrep** [comment]
I guess as it is released already, we will have to deal with it untill 5.2

**Comment by GrahamCampbell** [comment]
Yeh.

**Comment by GrahamCampbell** [comment]
5.2 will have a few breaking changes in it already too. Like, we're removing support for filters in the routing layer.

**Comment by TommyC81** [comment]
Maybe reopen this one?

**Comment by TommyC81** [comment]
@rkgrep @taylorotwell Maybe reopen this one? #10596

**Comment by GrahamCampbell** [comment]
No possible.

**Comment by GrahamCampbell** [comment]
1. The branch is deleted.
2. It's still not getting merged because it's to 5.1 rather than 5.2.

**Pull request closed without merging**